File: CallkitSample/CallkitSample/StringeeImplement/Models/CallkitCall.m
//
//  CallkitCall.m
//  CallkitSample
//
//  Created by HoangDuoc on 7/14/22.
//  Copyright © 2022 Hoang Duoc. All rights reserved.
//

#import "CallkitCall.h"
#import "WeakTimerTarget.h"
#import "CallManager.h"

@implementation CallkitCall {
    int counter;
    NSTimer *timeoutTimer;
}

- (instancetype)initWithIncomingState:(BOOL)isIncoming startTimer:(BOOL)startTimer {
    self = [super init];
    if (self) {
        self.isIncoming = isIncoming;
        if (startTimer) {
            [self startTimer];
        }
    }
    return self;
}

- (void)handleCallTimeOut {
    counter += 2;
    if (counter >= 28) {
        [self stopTimer];
        if (!self.answered && !self.rejected) {
            [[CallManager instance] endCall];
        }
    }
}

- (void)startTimer {
    if (timeoutTimer != nil) {
        return;
    }
    
    WeakTimerTarget *target = [[WeakTimerTarget alloc] initWithTarget:self selector:@selector(handleCallTimeOut)];
    timeoutTimer = [NSTimer scheduledTimerWithTimeInterval:2 target:target selector:@selector(timerDidFire:) userInfo:nil repeats:YES];
    [timeoutTimer fire];
}

- (void)stopTimer {
    [timeoutTimer invalidate];
    timeoutTimer = nil;
}

- (void)clean {
    [self stopTimer];
}

- (void)dealloc {
    [self stopTimer];
}


@end

File: CallkitSample-iOS13/CallkitSample-iOS13/StringeeImplement/Models/CallkitCall.swift
//
//  CallkitCall.swift
//  stringeex
//
//  Created by HoangDuoc on 12/31/19.
//  Copyright © 2019 HoangDuoc. All rights reserved.
//

import Foundation
import CallKit

class CallKitCall {
    var serial = 0
    var callId: String?
    var stringeeCall: StringeeCall?
    var uuid: UUID?
    
    var answered = false
    var rejected = false
    
    var audioIsActived = false
    var isIncoming = false
    var answerAction: CXAnswerCallAction?
    
    var timer: Timer?
    var counter = 0
    
    init(isIncoming: Bool, enableTimer: Bool = false) {
        self.isIncoming = isIncoming
        if enableTimer {
            startTimer()
        }
    }
    
    private func startTimer() {
        if timer != nil { return }
        
        stopTimer()
        timer = Timer(timeInterval: 2, target: self, selector: #selector(CallKitCall.handleCallTimeOut), userInfo: nil, repeats: true)
        RunLoop.current.add(timer!, forMode: .default)
        timer?.fire()
    }
    
    @objc private func handleCallTimeOut() {
        counter += 2
        
        if counter >= 28 {
            stopTimer()
            if !answered && !rejected {
                CallManager.shared.endCall()
            }
        }
    }
    
    private func stopTimer() {
        CFRunLoopStop(CFRunLoopGetCurrent())
        timer?.invalidate()
        timer = nil
    }
    
    func clean() {
        stopTimer()
    }
    
    deinit {
        stopTimer()
    }
}
